### PR TITLE
Scroller module

### PR DESCRIPTION
**Pull request by mixmix**
@ahdinosaur would be interested in talking about this.

my motivations were : 
- extract Scroller as a module
  - add styling to scroller, so that Message has border ONLY when it's a child of a Scroller
- hide some of the specific pickiness of the scroller setup.

what I've made is on the edge of not worth it ... but then the scroller code everywhere is a bit weird.

I think what I've done is a new pattern in a couple of ways : 
- I have a weird 'content' keyword to note the position of the content div to be created
- I'm introduced a meta-template / constructor which other templates are using...

### CODE REVIEW

**Comment by mixmix** [comment]
@pietgeursen what to review this ?

**Comment by dominictarr** [comment]
I should mention that since I wrote the scroller code I have thought that maybe it would be better to have a `state` object which was all the rows in the cache, instead of two streams.

**Comment by mixmix** [comment]
can you expand ?
By two streams, you're talking about the pull streams from pull-scroll
which are appending/ prepending based on scroll right?

By rows, are you refering to you original column and row classes ?
By cache are you meaning that global CACHE?

File: modules_basic/scroller.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+const fs = require('fs')
+const h = require('../h')
+
```
Comment (mixmix) [inline modules_basic/scroller.js]: @pietgeursen / @ahdinosaur this is the only file that really needs review

Comment (mixmix) [inline modules_basic/scroller.js]: then maybe look at one other file that implements the pattern. I'll post next to a good example

File: modules_basic/feed.js [hunk 2]
```diff
@@ -25,42 +26,33 @@ exports.create = function (api) {
...
-      var content = h('div.column.scroller__content')
-      var div = h('div.column.scroller',
-        {style: {'overflow':'auto'}},
-        h('div.scroller__wrapper',
-          h('div', api.avatar_profile(id)),
-          content
-        )
-      )
+    const profile = h('div', api.avatar_profile(id))
+    var { container, content } = api.build_scroller({ prepend: profile })
```
Comment (mixmix) [inline modules_basic/feed.js]: like here

**Comment by dominictarr** [comment]
sorry!

so currently, you create two streams that forever expand the scroll window. eventually this is too much, especially if you are doing initial sync. so, one thing, is when you scroll down lots, it should start to _remove_ items from the top, but that means, when you scroll back up, it needs to create a new pull-stream, from the last point that you have. Instead of passing two streams, you'd need to pass two stream constructors.

that would give you a growing/shrinking array (or maybe an `{}` with number keys, so you can go negative if you scroll way up?) maybe this is would be useful as a state object for choo/inu/react? the part that doesn't map nicely onto those frameworks, however, is that you sometimes need to manipulate the current scroll position so that the screen doesn't jump around.

**Comment by arj03** [comment]
In latest versions of patchbay changing tabs takes the tab jump to the top. Quite annoying for the main feed :) Does this fix that issue?

**Comment by dominictarr** [comment]
@arj03 you are right! that must have been introduced in the change to `visibility: hidden`?

**Comment by mixmix** [comment]
yeah I noticed that side effect too. Fixing bugs introducing bugs D:
This doesn't solve that.

**Comment by mixmix** [comment]
Yeah that ideal Scroller sounds awesome @dominictarr - I've [opened an issue](https://github.com/dominictarr/pull-scroll/issues/6) on pull-scroll to track this

Given I think that's future dev, what do people think of this as an intermediate step?

**Comment by mixmix** [comment]
@arj03 @dominictarr fix to scroll-position loss here : https://github.com/dominictarr/hypertabs/pull/5

File: modules_basic/scroller.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+exports.create = function (api) {
+  return {
+    build_scroller,
+    mcss: () => fs.readFileSync(__filename.replace(/js$/, 'mcss'), 'utf8')
+  }
+
+  function build_scroller ({ prepend = [], append = [] } = {}) {
+    var content = h('section.content')
```
Comment (pietgeursen) [inline modules_basic/scroller.js]: const

**Comment by pietgeursen** [comment]
standard style?

File: modules_basic/scroller.js (hunk 3 above)
Comment (mixmix) [inline modules_basic/scroller.js]: done

**Comment by mixmix** [comment]
thanks for the review @pietgeursen 
I've squashed this and am merging

**Pull request merged**